### PR TITLE
SLING-10018 Implement SelectionSet / SelectedField wrappers

### PR DESCRIPTION
**Pull request by tyge68**
- Defining 2 new interfaces (SelectedField and SelectionSet) to wrap GraphQL related DataFetching objects.
- Changed the SlingDataFetcherEnvironment to return the SelectionSet wrapper
- Changed the DataFetchingEnvironmentWrapper to wrap the SelectionSet in the returned wrapper.
- Added 2 implementing wrapper for SelectedFieldWrapper  and SelectionSetWrapper.
- Added a unit test

### CODE REVIEW

**Comment by tyge68** [comment]
@raducotescu  it would be nice if you could have a look at this suggested "add on" to the GraphQL Core. Let me know if you think that it would work, I would then check to add a few tests to cover it.

**Comment by raducotescu** [comment]
The API changes look ok to me. The only thing that I'd like to see in the API definitions are the JSR-305 annotations.

File: src/main/java/org/apache/sling/graphql/api/SelectedField.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+/**
+ * Interface to wrap GraphQL SelectedField.
+ */
+public interface SelectedField {
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/graphql/api/SelectedField.java:29]: The whole class needs a bit more details in its JavaDocs. In addition to that, I suspect that this will be a `@ProviderType`, right?

File: src/main/java/org/apache/sling/graphql/api/SelectionSet.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+/**
+ * Interface to wrap GraphQL DataFetchingFieldSelectionSet.
+ */
+public interface SelectionSet {
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/graphql/api/SelectionSet.java:29]: The whole class needs a bit more details in its JavaDocs. In addition to that, I suspect that this will be a `@ProviderType`, right?

File: src/main/java/org/apache/sling/graphql/core/engine/SelectedFieldWrapper.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        subFields = subFieldMap.values().stream().collect(Collectors.toList());
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public List<SelectedField> getSubSelectedFields() {
```
Comment (stefangrimm) [inline src/main/java/org/apache/sling/graphql/core/engine/SelectedFieldWrapper.java:70]: Typically we don't return `List`s directly (in AEM Sites). Not sure how Sling handles this.

If we stay with `List`, I would at least suggest to wrap the list with `Collections.unmodifiableList()`.

File: src/main/java/org/apache/sling/graphql/api/SelectionSet.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    @NotNull
+    List<SelectedField> getFields();
+
+    /**
+     * @return true if the field qualified name exist.
+     */
+    boolean contains(String qualifiedName);
+
+    /**
+     * @return SelectedField for qualified name.
```
Comment (stefangrimm) [inline src/main/java/org/apache/sling/graphql/api/SelectionSet.java:43]: The concept of "qualified name" should be explained (I don't want to have a look at the unit tests if I want to make use of an API, furthermore the unit tests only cover unions as a more complex use-case):

- How does a "qualified name" look like?
- How are things like unions and nested structures are reflected in the qualified name?

File: src/test/java/org/apache/sling/graphql/core/engine/DefaultQueryExecutorTest.java [hunk 5]
```diff
@@ -188,4 +191,30 @@ public void unionFetcherTest() throws Exception {
         assertThat(json, hasJsonPath("$.data.unionFetcher.items[0].testingArgument", equalTo("1, 2, 3")));
         assertThat(json, hasJsonPath("$.data.unionFetcher.items[1].path", equalTo(resource.getPath())));
     }
+
+    @Test
+    public void selectionSetTest() throws Exception {
```
Comment (stefangrimm) [inline src/test/java/org/apache/sling/graphql/core/engine/DefaultQueryExecutorTest.java:196]: This only covers unions. I would suggest to add a test for "normal" nested structures as well, as it would also be interesting how they would be reflected in the output, mostly if the type name would be included as well or if that's only true for `union`s. Also it would be interesting for the latter case to determine if an item of the selection set determines a union differentiator or actual data, as we couldn't simply follow the result data in the fetcher implementation in this case.

Comment (tyge68) [reply src/test/java/org/apache/sling/graphql/core/engine/DefaultQueryExecutorTest.java:196]: I will add a that in the test, only "inline" so currently in our cases Union would have the "type" as the name of the field.

**Comment by tyge68** [comment]
@raducotescu  @stefangrimm  I have added the ProviderType annotation, then improved the javadoc with more details, and added more cases to the test. Please let me know if that help to better understand this add on.

**Review by stefangrimm — COMMENTED** [review]
@raducotescu I'm fine with the implementation, although we might have missed some edge-cases (graphql-java is rather complex ...). But looks good for a first implementation.

@tyge68 I would suggest to (additionally) replace `private HashMap<...> field = new HashMap<>()` with `private Map<...> field = new HashMap<>()` to stick to interfaces for fields.
